feat(agent): add Web Search Agent using SerpAPI

royalpinto007 opened the pull request:
Introduces a Web Search Agent to perform web searches using SerpAPI with flexible query parameters and error handling. Future updates will include Jina Embeddings for Rankings and support for retrieving the latest videos from YouTube to streamline workflows.

Partial attempt to solve: https://github.com/video-db/Director/issues/61

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
	- Introduced a new Web Search Agent that performs searches using SerpAPI, specifically supporting video outputs.
	- Added functionality to handle video searches with structured responses.
	- Expanded the ChatHandler to include the new Web Search Agent.

- **Documentation**
	- Updated environment configuration to include a new variable for the SerpAPI integration.
  
- **Enhancements**
	- Improved handling of video IDs and agent prioritization within the reasoning engine for seamless workflows.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

royalpinto007 commented:
> @royalpinto007 can you please share a temp railway deployed version for testing?

For some reason, my deployed version on Railway wasn’t working as expected. I am currently trying to resolve the logs but am not sure what’s wrong. I can share the links here:

Frontend: https://frontend-production-8673.up.railway.app/
Backend: https://backend-production-5668.up.railway.app/

![Results Showcase](https://github.com/user-attachments/assets/0c0f0a7b-4229-4493-8853-a8fbad1c630f)

In the meantime, I can attach a screenshot of how things work here.

> Here's another good example of web search https://github.com/YassKhazzan/openperplex_backend_os

Regarding `openperplex_backend_os`, I did integrate it to check how it works, but it doesn’t seem to be efficient enough, at least for now, as it couldn’t even handle a simple search like, let’s say, “international news today.”

royalpinto007 commented:
@codeAshu 
The PR is good to go, as I believe the agent is performing as expected. I am attaching a screenshot to demonstrate how things are working.  

Unfortunately, the deployed version isn't functioning as expected, and the logs aren't providing much insight. I am not sure what am I missing- the backend has the following env variables- VIDEO_DB_API_KEY, SERP_API_KEY, while the frontend have these- VITE_APP_BACKEND_URL, VITE_PORT and VITE_OPEN_BROWSER. Though, I do see an error in the logs of the frontend deployment- Error: spawn xdg-open ENOENT. 

![1](https://github.com/user-attachments/assets/aaabb141-6390-406b-aa66-6e5830219276)
![2](https://github.com/user-attachments/assets/2e7401e2-8022-467c-b7b8-d1398512f838)
![3](https://github.com/user-attachments/assets/dda6e1d7-560c-4724-abb6-09d916df47ed)

codeAshu commented:
@ashish-spext and @0xrohitgarg Can you please review this and add your comments.

codeAshu commented:
@royalpinto007 we should show the search result in carousal to the user first when only `@websearch` is used by the user. 

You can borrow the same code we have to show collection of videos. This would require a PR in `videodb-chat` repo also. @0xrohitgarg can help you figure it out. You can DM on discord for this.

0xrohitgarg commented:
Hi @royalpinto007 👋,

I just finished testing this locally on my machine. This WebSearch Agent is going to super useful Add-On to the Director. Cheers.

Here's are some feedback where we can improve.

## Web Search Agent Enhancements

### SERP API Abstraction
- Abstract the SERP API under a tool to improve modularity and reusability.
- Checkout other tools examples [here](https://github.com/video-db/Director/tree/main/backend/director/tools) and how they are being used by agents.

### Video Picking Issues
- Picking up Videos
    - let's add only those videos which are either Youtube Links or Links which have `video_link` in the response
    - when its a non-youtube urls, the `link` key in video result contains a html page url to the blogpost/webpage of hosted video. ideally we want downloadable video url, there is a  `video_link` key in result, check that out too.
- Channel Selection Issue: Ensure it avoids picking up YouTube channels instead of individual videos.
### SERP API Parameters
- Clarify if duration filtering is supported in the SERP API (this isn't evident in the documentation).


## Frontend Updates

You can also show the Videos in a GridView that your search agent has picked up, so that user can preview them ( similar to perplexity behaviour)
We already have Frontend components in place, just follow these instructions to add the collection view to Your code.

- Pull changes from the [add-support-for-fal](https://github.com/video-db/Director/tree/add-support-for-fal) branch, it has neccessary backend Data types for List of Videos
- Run `make update-fe` to make sure your on latest version of [videodb-chat](https://github.com/video-db/videodb-chat)
- You can take inspiration from the [comparison_agent](https://github.com/video-db/Director/blob/add-support-for-fal/backend/director/agents/comparison.py), comparison agent is also rendering a list of videos on Frontend.

Add a loading state for videos
``` python   
videos_content = VideosContent(
    agent_name=self.agent_name,
    status=MsgStatus.progress,
    status_message="Searching for videos...",
    videos=[],
)

for i in range(count):
    video_data = VideoData(
        external_url="",
    )
    videos_content.videos.append(video_data)

self.output_message.content.append(videos_content)
self.output_message.push_update()
```

Then once you have video links available in your agents, just update `external_url` of specific index placeholder of that video from  `videos_content.videos`



## Architectural Considerations 

### Flexible Search Outputs 
> Think about this, can we create such schema that tommorow if we were to add support in our search agent to return text or images, we don't hvae to do many changes. 

- Enhance the web search agent to handle different types of searches based on the output typ
> We don't have to support all format right away, but schema can be designed in such a way that it is flexible to support these which doing many changes in schema, For example in [audio_generation](https://github.com/video-db/Director/blob/main/backend/director/agents/audio_generation.py) agent, we have different job types in agent that enable support of text_to_speech, sound_effects in single agent
    - Videos - currently supported
    - Text - furture
    - Images  - future

### Parameter Object Over Method

- Replace `Class.get_parameters` with an object for parameters.
- Pro: Parameters become reusable across agents (as done in the comparison agent).
- Con: Requires initializing the class every time parameters are needed.

ashish-spext commented:
Hi @royalpinto007 any updates? we are planning to close it this week.
Let us know if you need any additional information.

royalpinto007 commented:
> Hi @royalpinto007 any updates? we are planning to close it this week. Let us know if you need any additional information.

Sure, I am currently working on addressing the feedback provided by @0xrohitgarg and will share updates very soon. Sorry for the delay!

royalpinto007 commented:
Completed all updates. Can the PR be reviewed?

0xrohitgarg commented:
@royalpinto007 , can you please check the comments ?

0xrohitgarg commented on backend/director/tools/serp.py:
```diff
@@ -0,0 +1,80 @@
+                video_link = result.get("video_link", "")
+
+                # Skip channels or invalid links
+                if "channel" in link or "user" in link:
```
@royalpinto007 can you add more robust way of verifying the channel/user of youtube. ? 

just "channel" or "user" might also be present in other nonyt valid link

royalpinto007 commented on backend/director/tools/serp.py:
```diff
@@ -0,0 +1,80 @@
+        if duration:
+            if duration not in duration_mapping:
+                raise ValueError(f"Invalid duration value: {duration}")
+            params["tbs"] = duration_mapping[duration]  # Add duration filter
```
https://serpapi.com/blog/filtering-google-images-and-google-videos-results/

royalpinto007 commented:
@coderabbitai review

royalpinto007 commented:
@0xrohitgarg, could you please review it now?
